src/CachePlugin.php

```diff
@@ -67,12 +68,24 @@ public function handleRequest(RequestInterface $request, callable $next, callabl
 
         if ($cacheItem->isHit()) {
             // return cached response
-            return new FulfilledPromise($cacheItem->get());
+            $data = $cacheItem->get();
+            $response = $data['response'];
+            $response = $response->withBody($this->streamFactory->createStream($data['body']));
+
+            return new FulfilledPromise($response);
         }
 
         return $next($request)->then(function (ResponseInterface $response) use ($cacheItem) {
             if ($this->isCacheable($response)) {
-                $cacheItem->set($response)
+                $bodyStream = $response->getBody();
+                $body = $bodyStream->__toString();
+                if ($bodyStream->isSeekable()) {
+                    $bodyStream->rewind();
+                } else {
+                    $response = $response->withBody($this->streamFactory->createStream($body));
+                }
+
+                $cacheItem->set(['response' => $response, 'body' => $body])
```

Comment:
How awesome would it be if we could write a stream to the cache and the read it as a stream back from the cache. I guess it would require some kind of special backend which immediately limits our possibilities in the term of interoperability. It is still a good idea. :wink: